docs/website/docs/development/devfile.md

```diff
@@ -313,6 +313,182 @@ is invoked, that is, in this example, when the `run` or `deploy` commands are in
 
 Because the `kubernetes` component `k8s-deploybydefault-false-and-not-referenced` has `deployByDefault` set to `false` and is not referenced by any `apply` commands, it will never be applied.
 
+### How `odo` handles image names
+
+When the Devfile contains an Image Component with a relative `imageName` field, `odo` treats this field as an image name selector;
+it will scan the whole Devfile and automatically replace matching image names with a unique value that will be pushed to a user-defined registry.
+This replacement is done in matching Container components and manifests referenced in Kubernetes/OpenShift components.
+
+An image is said to match a relative image name if they both have the same name, regardless of their registry, tag, or digest.
+
+At the moment, `odo` only performs replacement in the manifests of the following core Kubernetes resources:
+- [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/)
+- [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
+- [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
+- [Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/)
+- [Pod](https://kubernetes.io/docs/concepts/workloads/pods/)
+- [ReplicaSet](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/)
+- [ReplicationController](https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller/)
+- [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+
+Replacement is done only if the user has set the `ImageRegistry` preference, which can be done via the `odo preference set ImageRegistry` command.
+See the [Configuration](../overview/configure.md#configuring-global-settings) page for more details.
+
+<details>
+<summary>Example</summary>
+
+```shell
+$ odo preference set ImageRegistry quay.io/$USER
+ ✓  Value of 'imageregistry' preference was set to 'quay.io/user'
+ ```
+
+</details>
+
+In this case, `odo` will automatically build relative Image components and push them to the specified registry path defined in the Preferences,
+using a dynamic image tag so that the images built and pushed can be unique across executions of `odo`.
+
+As an example, given the following Devfile excerpt (simplified to use only Kubernetes components, but the same behavior applies to OpenShift ones),
+and provided the `ImageRegistry` preference has been set (say to `quay.io/user`):
+
+```yaml
+schemaVersion: 2.2.0
+
+metadata:
+  name: my-app
+  version: 1.0.0
+
+components:
+  - image:
+      imageName: "my-relative-image:1.2.3-my-tag"
```

Comment:
This image name is supported, but is it the idiomatic way? (I cannot see any advantage for the user to add this tag, as it will never be used anywhere).

As a user, I would prefer to have documented first the idiomatic way to do, and have some notes to explain the corner cases.

Here, as a user, I would pass some time to try and understand what is the meaning of the tag, and as I understand, there is no meaning.

Reply:
Indeed. I wanted to show that the tag was ignored and replaced by a dynamic tag. But you're right; having a more realistic example would be better. I'll update it.